### ISSUE TITLE
Don't load select2 (and other JavaScript files) on all admin pages

### ISSUE THREAD

**Issue by sudar** [woothemes/woocommerce#8766]
WooCommerce is loading an [older version of select2](https://github.com/woothemes/woocommerce/pull/8123) for compatibility reason. But it is loading it in all admin pages which breaks other plugins that use v4.x of select2.

I am the author of Bulk Delete plugin and a I recently received a bug report from one of my users. Upon investigation I found that this is happening because WooCommerce is [loading select2 in all admin pages](https://github.com/sudar/bulk-delete/issues/111) (including my plugin's page) even though it doesn't need it.

My plugin is just one example, but there could be lot of other places where things might break because of this behaviour of WooCommerce. Is it possible to load WooCommerce related JavaScript files only on pages where it requires them?

**Comment by mikejolley** [comment]
It is registered, not enqueued.

https://github.com/woothemes/woocommerce/blob/master/includes/admin/class-wc-admin-assets.php#L100

it is a dependency of https://github.com/woothemes/woocommerce/blob/master/includes/admin/class-wc-admin-assets.php#L101

And this is only enqueued on WC screens https://github.com/woothemes/woocommerce/blob/master/includes/admin/class-wc-admin-assets.php#L126

Hope that makes sense - its not enqueued everywhere. I checked dashboard and there was no sign of it.

**Comment by sudar** [comment]
@mikejolley 

WooCommerce is infact enqueuing scripts on other pages as well. I have not found the exact part of WooCommerce code that is causing this, but can provide you the steps to replicate it.

- Install WooCommerce and my [Bulk Delete](https://wordpress.org/plugins/bulk-delete/) plugin on a test installation.
- Open any of my plugins custom page. Eg: Bulk WP -> Bulk Delete Posts
- You can see that WooCommerce enqueues its resources there.

Here is the screenshot of the view source page which shows that WooCommerce is enqueuing its assets on my plugin's custom page. You can see that it is my plugin page from the url.

<img width="1440" alt="screenshot 2015-08-15 08 51 16" src="https://cloud.githubusercontent.com/assets/103148/9287297/03edbd74-432b-11e5-8400-d9e66fbd352c.png">

Here is another screenshot of the same page scrolled a little further down that shows that WooCommerce is enqueuing select2 in my plugins page.

<img width="1440" alt="screenshot 2015-08-15 08 54 25" src="https://cloud.githubusercontent.com/assets/103148/9287301/4512f24c-432b-11e5-836a-f12bbf339436.png">
 
As I said, I have not gone over WooCommerce code in depth to understand why this is happening, but I am pretty sure WooCommerce is enqueuing its assets on non WooCommerce pages as well.

**Comment by sudar** [comment]
I did some more debugging and it seems that you are partially correct. WooCommerce is not enqueuing assets on non-WooCommerce pages.

But the problem is that you are registering your assets on all pages and in my script I have 'select2' as a dependency. Since you have already registered 'select2', WordPress is picking up WooCommerce's script rather than mine.

I have to explicitly deregister the select2 in my script and register mine so that I can use the correct version.

So would it be possible for you to move even the wp_register_script/styles call inside the condition that checks the current screen id with wc_get_screen_ids()?

**Comment by greguly** [comment]
Hi Sudar, 

If WooCommerce' s selct2 is a wrong version for your plugin, why don't you simply change your select2 name to bulk-delete-select2 ? 

That way you will not have trouble with any other plugins too.

Cheers,
Gabriel

**Comment by greguly** [comment]
BTW, nice plugin. I have https://wordpress.org/plugins/woocommerce-remove-all-products ;-)

**Comment by sudar** [comment]
@greguly 

> If WooCommerce' s selct2 is a wrong version for your plugin, why don't you simply change your select2 name to bulk-delete-select2 ?

Yes, I can do it. In fact I solved it by deregistering select2 registered by WooCommerce and then including the latest version.

The reason why I created this issue is not to just talk about just the specific issue that I had in my plugin but about the general issue that might affect any number of other plugins because of the way scripts are registered by WooCommerce.

For instance, once I figured out the actual issue, I was able to fix it in less time that it took me to create this issue, but what I am talking about is avoiding this kind of compatibility issues with not just my plugin but other plugins as well. Fixing compatibility issues is good, but avoiding compatibility issues is better :)

The other reason why using 'bulk-delete-select2' is a bad idea is because It is a vendor script that I am using without any modification. If all the plugins start to do this then there will be different copies of the same script enqueued in the same page which might cause other issues.

To give credit to WooCommerce they have infact handled most of the compatibility issues already. When I initially created this issue I thought WooCommerce was enqueueing all their assets on all the admin pages. But they are not doing it. They are only registering the assets and not enqueuing them and I was able to solve my issue only because of it. If they can also go one step above and even register their assets only on the pages where they need it, then it would be solve most of the potential problems.

> BTW, nice plugin. I have https://wordpress.org/plugins/woocommerce-remove-all-products ;-)

Glad to know that you like my plugin. Your plugin is nice as well :)
